### DISCUSSION TITLE
Changed source paths/filenames in 8.8.173 in combination with reportgenerator

### DISCUSSION THREAD

**Discussion by kimpenhaus** [SteveGilham/altcover#228] [Q&A] [answered]
Good morning @SteveGilham,

sorry bothering you again :) We've noticed a change in the generated cobertura files from 8.8.74 to 8.8.173 in the way the source elements and the filenames are generated.

I don't see an issue in this per se but with our setup the changes break the coverage analysis. we have a multi-project/assembly setup in azure that generats the code coverage with altcover in cobertura format and uses reportgenerator to "transform" this into single file cobertura and sonarqube generic format.

When updating to the latest version we realised that the reportgenerator threw lots of file not found warnings. drilling this down leeds us to the changes done by the new altcover verion. I will try to explain further:

In 8.8.74 it seems thet all filenames were absolute and the sources seem to contain all directories

assembly 1 (sample):

<img width="1850" alt="grafik" src="https://github.com/user-attachments/assets/7e576e6c-aebc-4c54-8300-adfcfefebdbe">
 
<img width="1668" alt="grafik" src="https://github.com/user-attachments/assets/9c0f70d5-83d2-4e2c-bbb3-6dc7f31ebf37">

assembly 2 (sample):

<img width="1456" alt="grafik" src="https://github.com/user-attachments/assets/ec6f8fc8-b501-4d8b-a4e3-315cf6f220c3">

<img width="1594" alt="grafik" src="https://github.com/user-attachments/assets/f803c85f-f7bb-496f-ab23-6c48abe2074f">

With the new 8.8.173 version same sources this seems to have changed in the way that sources contains the lowest common denominator (hopefully you understand what I mean with that :)) dircotryname and the filenames are relative to that:

assembly 1 (sample):

<img width="666" alt="grafik" src="https://github.com/user-attachments/assets/a5960e4e-9717-43f2-b8cb-985b15f37ee3">

<img width="1452" alt="grafik" src="https://github.com/user-attachments/assets/6f4cddcf-1ea0-480f-9269-15ce8479f145">

assembly 2 (sample):

<img width="1454" alt="grafik" src="https://github.com/user-attachments/assets/3382f723-f23a-4a21-adcb-f4978ec568d2">

<img width="474" alt="grafik" src="https://github.com/user-attachments/assets/4544944f-ed3a-4ff1-80bd-22b2472e8333">

the change itself looks conclusive and recduces the size of the generated files - so I don't see an issue here itself.

But when heading to reportgenerator we have the problem that the reportgenerator won't find these files.

What we tried is to set the `sourcedirs` parameter to `/agent/_work/1/s/src` but that will only fix those that matches with that directory in the sources. In the given samples `/StringExtensions.cs` would not be found as the filename is not relative to `/agent/_work/1/s/src`.

<img width="3208" alt="grafik" src="https://github.com/user-attachments/assets/36f06bc6-8b41-44e2-b149-dcefa89d7cc6">

as the assemblies (and maybe future assemblies) might vary in what sources will contain we have no idea what to set in the `sourcedirs` to cover this in general.

So our question is: Is there a way to have a root source directory which tells altcover to generate all sources and filename relative to that? So that we cann have `/agent/_work/1/s/src` in altcover and in reportgenerator?

Thanks for any suggestion 🙏🏼

**Comment by SteveGilham** [comment]
What you report here looks like there's a bug.

This change was intended as a resolution to issue #197, and also to bring the output into conformance with the original cobertura format, based as it is on the Java behaviour of keeping file path and full class name in step with one another.

It should be extracting longest common leading paths - in this case it should be extracting `/agent/_work/1/s/src`, or at worst `/agent` directly.

Background:

The class entry [in the cobertura source](https://github.com/cobertura/cobertura/blob/0ff963284cecaace30f409a977dccb07c41a5a8f/cobertura/src/main/java/net/sourceforge/cobertura/reporting/xml/XMLReport.java#L186) is like
```java
		println("<class name=\"" + classData.getName() + "\" filename=\""
				+ classData.getSourceFileName() + "\" line-rate=\""
				+ classData.getLineCoverageRate() + "\" branch-rate=\""
				+ classData.getBranchCoverageRate() + "\" complexity=\""
				+ complexity.getCCNForClass(classData) + "\"" + ">");
```
where [`getSourceFilename()`](https://github.com/cobertura/cobertura/blob/0ff963284cecaace30f409a977dccb07c41a5a8f/cobertura/src/main/java/net/sourceforge/cobertura/coveragedata/PackageData.java#L114) literally composes the path by replacing the separator characters in the class name
```java
	public String getSourceFileName() {
		return this.name.replace('.', '/');
	}
```
where it is assumed relative to one or other of the source directories.

**Comment by kimpenhaus** [comment]
Thanks for the quick reply and the insights

> It should be extracting longest common leading paths

I guess it does for each assembly independent (if I understood you correctly) but when handling all cobertura files together it does not - not sure if that is even possible?

**Comment by SteveGilham** [comment]
If this is a case of many assemblies being handled in the one coverage run, aggregation is possible, and it's an oversight on my part (due to only having tested on the single-assembly case) if it doesn't collate across all the assemblies. 

Collating multiple independent reports is a whole new requirement.

**Comment by kimpenhaus** [comment]
I am not pretty sure how `one coverage run` is defined - but I would assume we only have one run - but I am afraid I am not sure how azure handles this...

our pipeline task is as follows:

```
      - task: DotNetCoreCLI@2
        displayName: 'run unit tests'
        enabled: True
        inputs:
          command: 'test'
          projects: 'My.Integration.sln'
          arguments: '--nologo --no-build --configuration "$(BuildConfiguration)" /p:CopyLocalLockFileAssemblies=true /p:AltCover=true /p:AltCoverFailFast=true /p:AltCoverAssemblyFilter="^(?!My\.Integration).*$|^(My\.Integration.*?\.Test.*)$" /p:AltCoverAttributeFilter=ExcludeFromCodeCoverage /p:AltCoverFileFilter=\.g\.cs /p:AltCoverCobertura="coverage.cobertura.xml"'
```

**Comment by SteveGilham** [comment]
Alas, that's going to be multiple runs, one per test assembly.

I shall double-check what I have anyway, then consider how best to introduce a manually specified common root into the process.

**Reply by SteveGilham** [reply, marked as answer]
With release [v8.9.3](https://www.nuget.org/packages/altcover/8.9.3), the new `/p:AltCoverPackages=[list of source roots with '|' separation between entries]` allows you to specify one or more paths that are added to the existing selection of directories directly containing source, to be processed by the same longest leading common sub-tree approach.  

For your case, setting  `/p:AltCoverPackages=/agent/_work/1/s/src` should suffice, as that will then become the longest leading common sub-tree.
